File: lib/data/models/curso.dart
/// Trilha de aprendizagem. Espelha CursoDTO do backend.
class Curso {
  final int id;
  final String titulo;
  final String descricao;
  final String nivel; // BASICO | INTERMEDIARIO | AVANCADO
  final int cargaHoraria;
  final int totalModulos;
  int progresso; // 0-100 - mutavel: avanca conforme o usuario estuda
  final String? autorId; // nomeUser do tutor; null = curso oficial
  final String origem; // 'OFICIAL' | 'COMUNIDADE'
  final String status; // 'PUBLICADO' | 'RASCUNHO'
  final List<String> topicosModulos;

  Curso({
    required this.id,
    required this.titulo,
    required this.descricao,
    required this.nivel,
    required this.cargaHoraria,
    required this.totalModulos,
    this.progresso = 0,
    this.autorId,
    this.origem = 'OFICIAL',
    this.status = 'PUBLICADO',
    this.topicosModulos = const [],
  });

  bool get isComunidade => origem == 'COMUNIDADE';

  factory Curso.fromJson(Map<String, dynamic> json) => Curso(
        id: json['id'],
        titulo: json['titulo'] ?? '',
        descricao: json['descricao'] ?? '',
        nivel: json['nivel'] ?? 'BASICO',
        cargaHoraria: json['cargaHoraria'] ?? 0,
        totalModulos: json['totalModulos'] ?? 0,
        progresso: json['progresso'] ?? 0,
        autorId: json['autorId'],
        origem: json['origem'] ?? 'OFICIAL',
        status: json['status'] ?? 'PUBLICADO',
        topicosModulos: (json['topicosModulos'] as List<dynamic>?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
      );

  Map<String, dynamic> toJson() => {
        'titulo': titulo,
        'descricao': descricao,
        'nivel': nivel,
        'cargaHoraria': cargaHoraria,
        'topicosModulos': topicosModulos,
      };
}

File: backend-java/src/main/java/com/smarthas/backend/model/Curso.java
package com.smarthas.backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "cursos")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Curso {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String titulo;

    @Column(nullable = false, length = 500)
    private String descricao;

    // BASICO | INTERMEDIARIO | AVANCADO
    @Column(nullable = false)
    private String nivel;

    @Column(nullable = false)
    private Integer cargaHoraria;

    @Column(nullable = false)
    private Integer totalModulos;

    @Column(nullable = false)
    @Builder.Default
    private Integer progresso = 0;

    /** nomeUser do tutor; null = curso oficial ou tutor que excluiu a conta (orfanizado). */
    private Long autorId;

    // OFICIAL | COMUNIDADE
    @Column(nullable = false)
    @Builder.Default
    private String origem = "OFICIAL";

    // RASCUNHO | PUBLICADO
    @Column(nullable = false)
    @Builder.Default
    private String status = "PUBLICADO";

    // EAGER de proposito: colecao pequena (poucos topicos por curso) e este projeto
    // roda com spring.jpa.open-in-view=false, entao um fetch LAZY aqui quebraria a
    // serializacao JSON fora da transacao (LazyInitializationException).
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "curso_topicos_modulos", joinColumns = @JoinColumn(name = "curso_id"))
    @Column(name = "topico")
    @Builder.Default
    private List<String> topicosModulos = new ArrayList<>();
}
